### ISSUE TITLE
Metadata detection for Custom-Aggregations unreliable

### ISSUE THREAD

**Issue by chmp** [dask/dask#2708]
In `.groupby(...).agg(...)`, first a temporary is created via `aca`, which is then passed through `map_partitions`. Thereby, the metadata of a the aggregated dataframe is determined via the builtin mechanism of `map_partitions`. This implementation will fail in case objects are used as intermediate results. 

The motivation of using object intermediates can be to hide the values from pandas. For example, the mode function could be implemented as

```python
def chunk_mode(s):
    return s.apply(lambda s: [s.value_counts()])


def agg_mode(s):
    def impl(s):
        res, = s.iloc[0]
        
        for i, in s.iloc[1:]:
            res = res.add(i, fill_value=0)
        
        return [res]
    
    return s.apply(impl)
    

def finalize_mode(s):
    # fake result to work around dask metadata mechanism 
    if not isinstance(s.iloc[0], list):
        return s
    
    return s.map(lambda i: i[0].argmax())
```

However, as the result-column after `aca` will be of object-type the corresponding meta object will be filled with strings, whereas the true column will be filled with lists of series. Here, I used a type check to detect whether the code is executed during meta-data detection or during the true aggregation.

A preferable way would be to re-implement the meta-data detection code for `.agg` to not use intermediate dask dataframes, but to apply the full sequence of `chunk, agg, finalize` steps to transform the input metadata to output  metadata.

For the next weeks, I wont have time to work on this issue myself, but I wanted to document the behavior in case someone else stumbles over it or wants to improve the situation themselves.

For reference the traceback with `dask==0.15.3`, if the workaround is removed:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname)
    136     try:
--> 137         yield
    138     except Exception as e:

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   3127     with raise_on_meta_error(funcname(func)):
-> 3128         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   3129 

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py in _agg_finalize(df, funcs)
    637     for result_column, func, kwargs in funcs:
--> 638         result[result_column] = func(df, **kwargs)
    639 

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py in _apply_func_to_columns(df_like, prefix, func)
    659     columns = [df_like[col] for col in columns]
--> 660     return func(*columns)
    661 

<ipython-input-3-29da7874ad98> in <lambda>(s)
     15     agg_mode,
---> 16     lambda s: s.map(lambda i: i[0].argmax()),
     17 )

~/venvs/py3-data-science/lib/python3.6/site-packages/pandas/core/series.py in map(self, arg, na_action)
   2157             # arg is a function
-> 2158             new_values = map_f(values, arg)
   2159 

pandas/_libs/src/inference.pyx in pandas._libs.lib.map_infer (pandas/_libs/lib.c:66645)()

<ipython-input-3-29da7874ad98> in <lambda>(i)
     15     agg_mode,
---> 16     lambda s: s.map(lambda i: i[0].argmax()),
     17 )

AttributeError: 'str' object has no attribute 'argmax'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-3-29da7874ad98> in <module>()
     24 a = dd.from_pandas(d, npartitions=5)
     25 
---> 26 actual = a['cc'].groupby([a['g0'], a['g1']]).agg(agg_func)
     27 
     28 # cheat to get the correct index

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py in agg(self, arg, split_every, split_out)
   1189     @derived_from(pd.core.groupby.SeriesGroupBy)
   1190     def agg(self, arg, split_every=None, split_out=1):
-> 1191         return self.aggregate(arg, split_every=split_every, split_out=split_out)

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1178     @derived_from(pd.core.groupby.SeriesGroupBy)
   1179     def aggregate(self, arg, split_every=None, split_out=1):
-> 1180         result = super(SeriesGroupBy, self).aggregate(arg, split_every=split_every, split_out=split_out)
   1181         if self._slice:
   1182             result = result[self._slice]

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1008 
   1009         return map_partitions(_agg_finalize, obj, token='aggregate-finalize',
-> 1010                               funcs=finalizers)
   1011 
   1012     @insert_meta_param_description(pad=12)

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/core.py in map_partitions(func, *args, **kwargs)
   3160 
   3161     if meta is no_default:
-> 3162         meta = _emulate(func, *args, **kwargs)
   3163 
   3164     if all(isinstance(arg, Scalar) for arg in args):

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   3126     """
   3127     with raise_on_meta_error(funcname(func)):
-> 3128         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   3129 
   3130 

/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     98                 value = type()
     99             try:
--> 100                 self.gen.throw(type, value, traceback)
    101                 raise RuntimeError("generator didn't stop after throw()")
    102             except StopIteration as exc:

~/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname)
    148                ).format(" in `{0}`".format(funcname) if funcname else "",
    149                         repr(e), tb)
--> 150         raise ValueError(msg)
    151 
    152 

ValueError: Metadata inference failed in `_agg_finalize`.

Original error is below:
------------------------
AttributeError("'str' object has no attribute 'argmax'",)

Traceback:
---------
  File "/Users/private/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/utils.py", line 137, in raise_on_meta_error
    yield
  File "/Users/private/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/core.py", line 3128, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/Users/private/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py", line 638, in _agg_finalize
    result[result_column] = func(df, **kwargs)
  File "/Users/private/venvs/py3-data-science/lib/python3.6/site-packages/dask/dataframe/groupby.py", line 660, in _apply_func_to_columns
    return func(*columns)
  File "<ipython-input-3-29da7874ad98>", line 16, in <lambda>
    lambda s: s.map(lambda i: i[0].argmax()),
  File "/Users/private/venvs/py3-data-science/lib/python3.6/site-packages/pandas/core/series.py", line 2158, in map
    new_values = map_f(values, arg)
  File "pandas/_libs/src/inference.pyx", line 1574, in pandas._libs.lib.map_infer (pandas/_libs/lib.c:66645)
  File "<ipython-input-3-29da7874ad98>", line 16, in <lambda>
    lambda s: s.map(lambda i: i[0].argmax()),
```

**Comment by chmp** [comment]
Ok, that was easier than I initially thought. I pushed a fix as #2712.

**Comment by chmp** [comment]
Closed via #2712.
